driver-core/src/test/java/com/datastax/driver/core/ControlConnectionTest.java

```diff
@@ -560,6 +571,292 @@ public void should_connect_when_peers_v2_table_not_present() {
+             * both peer and rpc_address */
+            .peers("peer", InetAddress.getByName("1.2.3.4"))
+            .expectedAddress(expectedAddress)
+            .build();
+    runPeerTest(state);
+  }
+
+  private void runPeerTest(PeerRowState state) {
+
+    ScassandraCluster scassandras =
+        ScassandraCluster.builder().withNodes(2).withPeersV2(state.usePeersV2()).build();
+    scassandras.init();
+
+    Cluster cluster = null;
+    try {
+
+      scassandras.node(1).primingClient().clearAllPrimes();
+
+      PrimingClient primingClient = scassandras.node(1).primingClient();
+
+      /* Note that we always prime system.local; ControlConnection.refreshNodeAndTokenMap() gets angry
+       * if this is empty */
+      primingClient.prime(
+          PrimingRequest.queryBuilder()
+              .withQuery("SELECT * FROM system.local WHERE key='local'")
+              .withThen(then().withColumnTypes(SELECT_LOCAL).withRows(state.getLocalRow()).build())
+              .build());
+
+      if (state.shouldPrimePeers()) {
+
+        primingClient.prime(
+            PrimingRequest.queryBuilder()
+                .withQuery("SELECT * FROM system.peers")
+                .withThen(
+                    then()
+                        .withColumnTypes(state.isDse68() ? SELECT_PEERS_DSE68 : SELECT_PEERS)
+                        .withRows(state.getPeersRow())
+                        .build())
+                .build());
+      }
+      if (state.shouldPrimePeersV2()) {
+
+        primingClient.prime(
+            PrimingRequest.queryBuilder()
+                .withQuery("SELECT * FROM system.peers_v2")
+                .withThen(
+                    then().withColumnTypes(SELECT_PEERS_V2).withRows(state.getPeersV2Row()).build())
+                .build());
+      } else {
+
+        /* Must return an error code in this case in order to trigger the driver's downgrade to system.peers */
+        primingClient.prime(
+            PrimingRequest.queryBuilder()
+                .withQuery("SELECT * FROM system.peers_v2")
+                .withThen(then().withResult(Result.invalid).build()));
+      }
+
+      cluster =
+          Cluster.builder()
+              .addContactPoints(scassandras.address(1).getAddress())
+              .withPort(scassandras.getBinaryPort())
+              .withNettyOptions(nonQuietClusterCloseOptions)
+              .build();
+      cluster.connect();
+
+      Collection<EndPoint> hostEndPoints =
+          Collections2.transform(
+              cluster.getMetadata().allHosts(),
+              new Function<Host, EndPoint>() {
+                public EndPoint apply(Host host) {
+                  return host.getEndPoint();
+                }
+              });
+      assertThat(hostEndPoints).contains(state.getExpectedEndPoint(scassandras));
+    } finally {
+      if (cluster != null) cluster.close();
+      scassandras.stop();
+    }
+  }
+
+  static class PeerRowState {
```

Comment:
State object + builder is a bit more convoluted than I'd like but it has the effect of nicely isolating changes required for an individual test.  Without this class I found myself passing around multiple config options (all of which were either related or similar to each other) in order to setup a given test.  Introducing this class had the effect of making the intent of a single test case above very clear.